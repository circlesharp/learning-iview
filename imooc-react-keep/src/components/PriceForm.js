import React from 'react';
import { isValidDate } from '../utility';

class PriceForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      validatePass: true,
      errorMessage: '',
    };
  }

  submitForm = event => {
    const { item, onFormSubmit } = this.props;
    const editMode = item && !!item.id;
    const title = this.titleInput.value.trim();
    const price = + this.priceInput.value.trim();
    const date = this.dateInput.value.trim();

    event.preventDefault();

    if (title && price && date) {
      if (price < 0) {
        this.setState({
          validatePass: false,
          errorMessage: '价格数字必须大于0',
        });
      } else if (!isValidDate(date)) {
        console.log(date, 123);
        this.setState({
          validatePass: false,
          errorMessage: '请填写正确的日期格式',
        });
      } else {
        this.setState({
          validatePass: true,
          errorMessage: '',
        });

        if (editMode)
          onFormSubmit({ ...item, title, price, date }, true);
        else
          onFormSubmit({ title, price, date }, false);
      }
    } else {
      this.setState({
        validatePass: false,
        errorMessage: '请输入所有必选项',
      });
    }
  };

  render() {
    const { title, price, date } = this.props.item || {};

    return (
      <form
        className="price-form-component"
        onSubmit={event => this.submitForm(event)}
        noValidate
      >
        {/* 1 标题 */}
        <div className="form-group">
          <label htmlFor="title">标题 *</label>
          <input
            id="title"
            type="text"
            className="form-control"
            defaultValue={title}
            ref={titleInput => { this.titleInput = titleInput; }}
          />
        </div>

        {/* 2 价格 */}
        <div className="form-group">
          <label htmlFor="price">价格 *</label>
          <div className="input-group">
            <div className="input-group-prepend">
              <span className="input-group-text">￥</span>
            </div>
          </div>
          <input
            id="price"
            type="number"
            className="form-control"
            defaultValue={price}
            placeholde="请输入价格"
            ref={priceInput => { this.priceInput = priceInput; }}
          />
        </div>

        {/* 3 日期 */}
        <div className="form-group">
          <label htmlFor="date">日期 *</label>
          <input
            id="date"
            type="date"
            className="form-control"
            placeholder="请输入日期"
            defaultValue={date}
            ref={dateInput => { this.dateInput = dateInput; }}
          />
        </div>

        {/* 4 按钮组 */}
        <button type="submit" className="btn btn-primary mr-3">提交</button>
        <button
          className="btn btn-secondary"
          onClick={this.props.onCancelSubmit}
        >取消</button>

        {/* 5 alert */}
        {
          !this.state.validatePass &&
          <div className="alert alert-danger mt-5" role="alert">
            {this.state.errorMessage}
          </div>
        }
      </form>
    );
  }
}

export default PriceForm;
