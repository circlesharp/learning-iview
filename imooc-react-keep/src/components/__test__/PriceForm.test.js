import React from 'react';
import { mount } from 'enzyme';
import PriceForm from '../PriceForm';
import { testItems } from '../../testData';

let props = {
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn(),
};

let props_with_item = {
  item: testItems[0],
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn(),
};

let wrapper, wrapper2, formInstance, formInstance2;

export const getInputValue = (selector, wrapper) =>
  wrapper.find(selector).instance().value;

export const setInputValue = (selector, newValue, wrapper) => {
  wrapper.find(selector).instance().value = newValue;
};

describe('test PriceForm component', () => {
  beforeEach(() => {
    wrapper = mount(<PriceForm {...props} />);
    wrapper2 = mount(<PriceForm {...props_with_item} />);
    formInstance = wrapper.find(PriceForm).instance();
    formInstance2 = wrapper2.find(PriceForm).instance();
  });

  /* 1 snapshot */
  it('snapshot', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper2).toMatchSnapshot();
  });

  /* 2 Price with no item */
  describe('Price with no item', () => {
    /* 含 3 inputs, 1 form */
    it('含 3 inputs, 1 form', () => {
      expect(wrapper.find('input').length).toEqual(3);
      expect(wrapper.find('form').length).toEqual(1);
    });

    /* 默认表单空 */
    it('默认表单空', () => {
      expect(getInputValue('#title', wrapper)).toEqual('');
      expect(getInputValue('#price', wrapper)).toEqual('');
      expect(getInputValue('#date', wrapper)).toEqual('');
    });

    /* 提交空表单 */
    it('提交空表单', () => {
      wrapper.find('form').simulate('submit');

      expect(formInstance.state.validatePass).toEqual(false);
      expect(wrapper.find('.alert').length).toEqual(1);
      expect(props.onFormSubmit).not.toHaveBeenCalled(); // ??
    });

    /* 提交不合法的价格 */
    it('提交不合法的价格', () => {
      setInputValue('#title', 'test', wrapper);
      setInputValue('#price', '-20', wrapper);
      wrapper.find('form').simulate('submit');
      expect(formInstance.state.validatePass).toEqual(false);
    });

    /* 提交不合法的日期 */
    it('提交不合法的日期', () => {
      setInputValue('#title', 'test', wrapper);
      setInputValue('#price', '20', wrapper);
      setInputValue('#date', 'wrong date', wrapper);
      wrapper.find('form').simulate('submit');
      expect(formInstance.state.validatePass).toEqual(false);
    });

    /* 表单校验通过并触发提交回调 */
    it('表单校验通过并触发提交回调', () => {
      const propsItem = testItems[1];
      const emitItem = { title: '这是我的工资', price: 20000, date: '2018-08-18' };
      setInputValue('#title', propsItem.title, wrapper);
      setInputValue('#price', propsItem.price, wrapper);
      setInputValue('#date', propsItem.date, wrapper);
      wrapper.find('form').simulate('submit');
      expect(props.onFormSubmit).toHaveBeenCalledWith(emitItem, false);
    });

    /* 触发取消回调 */
    it('触发取消回调', () => {
      wrapper.find('button').last().simulate('click');
      expect(props.onCancelSubmit).toHaveBeenCalled();
    });
  });

  /* 3 测试带有初始值 */
  describe('test PriceForm with item date', () => {
    /* 显示初始值 */
    it('显示初始值', () => {
      expect(getInputValue('#title', wrapper2)).toEqual(props_with_item.item.title);
      expect(getInputValue('#price', wrapper2)).toEqual(props_with_item.item.price.toString());
      expect(getInputValue('#date', wrapper2)).toEqual(props_with_item.item.date);
    });

    it('修改默认值', () => {
      const emitItem = { title: '这是我的工资', price: 20000, date: '2018-08-18' };
      setInputValue('#title', emitItem.title, wrapper2);
      setInputValue('#price', emitItem.price, wrapper2);
      setInputValue('#date', emitItem.date, wrapper2);
      wrapper2.find('form').simulate('submit');

      expect(formInstance2.state.validatePass).toEqual(true);
      expect(wrapper2.find('.alert').length).toEqual(0);
      expect(props_with_item.onFormSubmit).toHaveBeenCalledWith({ ...props_with_item.item, ...emitItem }, true);
    });
  });
});
