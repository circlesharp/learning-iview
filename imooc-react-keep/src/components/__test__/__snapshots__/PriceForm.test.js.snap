// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceForm component snapshot 1`] = `ReactWrapper {}`;

exports[`test PriceForm component snapshot 2`] = `ReactWrapper {}`;
